FWCore/Framework/interface/global/outputmoduleAbilityToImplementor.h

```diff
@@ -48,6 +48,37 @@ namespace edm {
         virtual void respondToCloseInputFile(FileBlock const&) = 0;
       };
 
+      template <typename T, typename C>
+      class StreamCacheHolder : public virtual T {
+      public:
+        explicit StreamCacheHolder(edm::ParameterSet const& iPSet) : OutputModuleBase(iPSet) {}
+        StreamCacheHolder(StreamCacheHolder<T, C> const&) = delete;
+        StreamCacheHolder<T, C>& operator=(StreamCacheHolder<T, C> const&) = delete;
+        ~StreamCacheHolder() override {
+          for (auto c : caches_) {
+            delete c;
+          }
+        }
+
+      protected:
+        C* streamCache(edm::StreamID iID) const { return caches_[iID.value()]; }
+
+      private:
+        void preallocStreams(unsigned int iNStreams) final { caches_.resize(iNStreams, static_cast<C*>(nullptr)); }
+        void doBeginStream_(StreamID id) final { caches_[id.value()] = beginStream(id).release(); }
+        void doEndStream_(StreamID id) final {
+          endStream(id);
+          delete caches_[id.value()];
+          caches_[id.value()] = nullptr;
+        }
+
+        virtual std::unique_ptr<C> beginStream(edm::StreamID) const = 0;
+        virtual void endStream(edm::StreamID) const {}
+
+        //When threaded we will have a container for N items whre N is # of streams
+        std::vector<C*> caches_;
```

Comment:
Just curious (since the pattern is already elsewhere), why are we not using `std::vector<std::unique_ptr<C>>` here?

Reply:
good question.